crates/pathfinder/src/bin/pathfinder/main.rs

```diff
@@ -53,35 +53,46 @@ async fn main() -> anyhow::Result<()> {
             .context("Starting monitoring task")?;
     }
 
-    let ethereum = EthereumContext::setup(config.ethereum.url, config.ethereum.password)
-        .await
-        .context("Creating Ethereum context")?;
-
-    // Use the default starknet network if none was configured.
+    let ethereum_client = if let Some(password) = config.ethereum.password {
+        EthereumClient::new_with_password(config.ethereum.url, &password)?
+    } else {
+        EthereumClient::new(config.ethereum.url)
+    };
+    let ethereum_chain_id = ethereum_client.chain_id().await?;
     let network = match config.network {
-        Some(network) => network,
-        None => ethereum
-            .default_network()
-            .context("Using default Starknet network based on Ethereum configuration")?,
+        Some(network) => match (&network, &ethereum_chain_id) {
+            (NetworkConfig::Mainnet, EthereumChain::Mainnet) => network,
+            (NetworkConfig::Testnet, EthereumChain::Goerli) => network,
+            (NetworkConfig::Testnet2, EthereumChain::Goerli) => network,
+            (NetworkConfig::Integration, EthereumChain::Goerli) => network,
+            (NetworkConfig::Custom { .. }, _) => network,
+            _ => anyhow::bail!(
+                "Network setup mismatch: Starknet={}, Ethereum={}",
+                network,
+                ethereum_chain_id
+            ),
+        },
+        None => match ethereum_chain_id {
+            EthereumChain::Mainnet => NetworkConfig::Mainnet,
+            EthereumChain::Goerli => NetworkConfig::Testnet,
+            EthereumChain::Other(id) => {
+                anyhow::bail!("Network is undefined. Ethereum chain_id={id}")
+            }
+        },
```

Comment:
Can we move this back into a separate function? Originally the entire main file looked like this -- with logic fully expanded. Made it very difficult to grok what was happening.

Reply:
Done